[reetmitra] iP

reetmitra opened the pull request:
# EchoPilot

> "Simplify your digital life with automation." - EchoPilot

**EchoPilot** is designed to be:
- 🚀 **Fast** and efficient, simplifying task management.
- 📚 **User-friendly**, with a text-based UI for easy navigation.
- 🌟 **Free** and open-source, enabling customization and contributions.

Getting started is as easy as:

1. **Download EchoPilot**:
   - Navigate to the [Releases](https://github.com/reetmitra/ip/releases) section of EchoPilot's GitHub repository.
   - Download the `duke.jar` file from the latest release.
2. **Run EchoPilot**:
   - Open a terminal or command prompt.
   - Navigate to the directory where you downloaded `duke.jar`.
   - Execute the following command:
     ```shell
     java -jar duke.jar
     ```
   - EchoPilot will start, and you can begin interacting with it through the command line interface.

## Features:

- Managing daily tasks
- Custom reminders 📅
- Seamless integration with your workflow

For developers 🚀, EchoPilot offers a great opportunity to dive into automation programming. Here's a peek at the `main` method in Java:

```java
    public static void main(String[] args) {
        new Duke("./data/duke.txt").run();
    }
```
### Roadmap:

- [x] Basic task management
- [ ] Advanced natural language processing
- [ ] Integration with third-party services

Your contributions and feedback are welcome! Don't forget to ⭐️ *star* ⭐️ the project if you find EchoPilot helpful.

"_Automate your routine, focus on what truly matters._" - EchoPilot

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

Commit:
docs/README.md: Tweak document template

Let's tweak the docs/README.md (which is used as the user guide)
to fit Duke better. Specifically,

1. mention product name in the title
2. mention adding a product screenshot and a product intro
3. tweak the flow to describe feature-by-feature

reetmitra committed:
Rename, Greet, Exit

reetmitra committed:
Echo

reetmitra committed:
Add, List

reetmitra committed:
Mark as Done

reetmitra committed:
ToDos, Events, Deadlines

reetmitra committed:
Automated Text UI Testing

reetmitra committed:
Handle Errors

reetmitra committed:
Delete

reetmitra committed:
Level 7 Saving implemented

reetmitra committed:
data file

reetmitra committed:
Edit some parts to match expected outputs

reetmitra committed:
Dates and Times

reetmitra committed:
Refactor code to match A-MoreOOP properties

reetmitra committed:
Organize the classes into suitable packages

reetmitra committed:
Using Gradle to automate build tasks.

reetmitra committed:
Use Gradle to automate some build tasks: Set up gradle to build and run Duke.

0-yibai commented on src/main/java/duke/ui/Ui.java:
```diff
@@ -0,0 +1,46 @@
+                + message + "\n"
+                + "____________________________________________________________";
+        System.out.println(errorMessage);
+    }
```
showError() and showMessage() methods looks the same to me. Maybe add a comment to explain why having them as 2 separate methods.

0-yibai commented on src/main/java/duke/task/Event.java:
```diff
@@ -0,0 +1,23 @@
+
+public class Event extends Task {
+    protected String from;
+    protected String to;
```
Perhaps it's better to name `from` as `startTime` or `timeFrom` and `to` as `endTime` or `timeTo` to reduce ambiguity in meaning?

DingBao-sys replied:
I agree with yibai.

0-yibai commented on src/main/java/duke/parser/Parser.java:
```diff
@@ -0,0 +1,82 @@
+
+        if (newTask != null) {
+            tasks.addTask(newTask);
+            ui.showMessage("Got it. I've added this task:\n  " + newTask + "\nNow you have " + tasks.getTasks().size() + " tasks in the list.");
```
Would using `String.format(...)` improve readability here?

0-yibai commented on src/main/java/duke/task/Event.java:
```diff
@@ -0,0 +1,23 @@
+
+    @Override
+    public String toString() {
+        return "E | " + (isComplete ? 1 : 0) + " | " + description + " | " + from + " to " + to;
```
Perhaps could make use of polymorphism here so that `isComplete` and `description` can be made `private` rather than `protected` for better encapsulation, i.e. call `super.toString()` here. Same for `toFileFormat()`.

DingBao-sys replied:
I think since the task are similar in terms of their file format, as said by YiBai , I think you could have Task handle the similarities and then use inheritance to add the details specific to the various task types

```suggestion
        public class Task {
           /**
              some implementation of filefomrat
           */
        }
        
        public class Event extends Task {
            /**
                call parent file format with super.fileformat()
            */
        }
```

0-yibai commented on src/main/java/duke/task/Task.java:
```diff
@@ -0,0 +1,32 @@
+
+public abstract class Task {
+    protected String description;
+    protected boolean isComplete;
```
I like your naming of boolean because it conforms to code quality guidelines 😺

0-yibai commented on src/main/java/duke/task/TaskList.java:
```diff
@@ -0,0 +1,59 @@
+import java.util.List;
+
+public class TaskList {
+    private List<Task> tasks;
```
I like your naming of list objects as plural because it conforms to code quality guidelines 😺

reetmitra committed:
Add JUnit tests to test the behavior of the code

0-yibai reviewed (commented):
LGTM. Overall clean and clear code. Here are some minor suggestions on variable naming and OOP.

reetmitra committed:
Data file changed

kinjalagarwal1810 commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,45 @@
+import duke.storage.Storage;
+import duke.task.TaskList;
+import duke.parser.Parser;
+
```
You may want to add javadocs for better readability.

kinjalagarwal1810 commented on src/main/java/duke/parser/Parser.java:
```diff
@@ -0,0 +1,82 @@
+        String command = parts[0].toLowerCase();
+
+        switch (command) {
+            case "bye":
```
Case should be in line with "switch". No extra indentation required.

kinjalagarwal1810 commented on src/main/java/duke/parser/Parser.java:
```diff
@@ -0,0 +1,82 @@
+
+        Task newTask = null;
+        switch (command) {
+            case "todo":
```
Case should be in line with "switch". No extra indentation required.

kinjalagarwal1810 commented on src/main/java/duke/storage/Storage.java:
```diff
@@ -0,0 +1,67 @@
+                switch (parts[0]) {
+                    case "T":
+                        task = new Todo(parts[2], parts[1].equals("1"));
+                        break;
```
Perhaps intead of using "parts" you can be more specific as to what it refers to e.g. "eventDetails" , "deadlineDetails", "eventTiming"

reetmitra committed:
Package app as JAR file for easier distribution

kinjalagarwal1810 commented on src/main/java/duke/task/Task.java:
```diff
@@ -0,0 +1,32 @@
+
+    public String getStatusIcon() {
+        return (isComplete ? "X" : " ");
+    }
```
Good variable names! Easy to understand

kinjalagarwal1810 reviewed (commented):
Looks good! Very well written code that is easy to follow

reetmitra committed:
Add JavaDoc comments to the code.

reetmitra committed:
Add find functionality to allow users to find a task by searching keyword

DingBao-sys commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,60 @@
+import duke.parser.Parser;
+
+public class Duke {
+    private Storage storage;
+    private TaskList tasks;
+    private Ui ui;
```
I like how you use very self explanatory names

DingBao-sys commented on src/main/java/duke/parser/Parser.java:
```diff
@@ -0,0 +1,105 @@
+                tasks.findTask(parts[1]);
+                break;
+            default:
+                throw new DukeException("I'm sorry, but I don't know what that means :-(");
```
perhaps you can create an Exception that extends DukeException, and call it UnknownInputException so that it is clear

DingBao-sys commented on src/main/java/duke/parser/Parser.java:
```diff
@@ -0,0 +1,105 @@
+
+        if (newTask != null) {
+            tasks.addTask(newTask);
+            ui.showMessage("Got it. I've added this task:\n  " + newTask + "\nNow you have " + tasks.getTasks().size() + " tasks in the list.");
```
Maybe some line breaks to improve readibility

DingBao-sys commented on src/main/java/duke/storage/Storage.java:
```diff
@@ -0,0 +1,84 @@
+     * @param tasks The list of tasks to be saved.
+     * @throws DukeException If an error occurs while saving tasks to the file.
+     */
+    public void save(List<Task> tasks) throws DukeException {
+        try {
+            File directory = new File(filePath).getParentFile();
+            if (!directory.exists()) {
+                directory.mkdirs();
+            }
+            FileWriter writer = new FileWriter(filePath);
+            for (Task task : tasks) {
+                writer.write(task.toFileFormat() + "\n");
+            }
+            writer.close();
+        } catch (IOException e) {
+            throw new DukeException("An error occurred while saving tasks: " + e.getMessage());
+        }
+    }
+}
```
maybe you can rename the function to saveToDrive so that it will be clear that you are saving the task to the computer drive and not on the data structure used during the program

DingBao-sys reviewed (commented):
Overall I think good ip done and good adherence to the coding standards

reetmitra committed:
Add checkstyle functionality

reetmitra committed:
Edit code to match CheckStyle coding standards

reetmitra committed:
Implement initial version of GUI interface

reetmitra committed:
Implement GUI for task management application

This commit includes the implementation of a basic graphical user interface for the task management application.

reetmitra committed:
Implement GUI for task management application

This commit includes the implementation of a basic graphical user interface for the task management application.

reetmitra committed:
Add Assertions for Data Validation

- Ensure non-empty task descriptions
- Validate successful date parsing in Deadline
- Check for redundant task completions
- Assert task index within bounds for retrieval

reetmitra committed:
Added Level-10 increment

reetmitra committed:
Standardize Full Commit Messages

Enforce a consistent commit message format with a concise title and a detailed body. This improves project history readability and aids in tracking changes effectively.

- Set character limits for titles and bodies
- Categorize commits for easy navigation
- Aim for clearer project evolution understanding

This change streamlines collaboration and project maintenance.

reetmitra committed:
Incorporate Assertions for Integrity Checks

- Verify task descriptions are not blank.
- Confirm correct date conversion in Deadline tasks.
- Prevent duplicate task completion actions.
- Safeguard against out-of-range task index access.

reetmitra committed:
Enhance Code Quality and Maintainability
Refactored code for improved readability and efficiency. Applied best practices and standardized coding conventions to ensure long-term maintainability and scalability of the project. Enhanced documentation for clearer understanding and future development ease.Incorporate Assertions for Integrity Checks

- Verify task descriptions are not blank.
- Confirm correct date conversion in Deadline tasks.
- Prevent duplicate task completion actions.
- Safeguard against out-of-range task index access.

reetmitra committed:
Add B-DoWithinPeriodTasks extension

This commit adds support for managing tasks that need to be done within
a specified period. The DoWithinPeriodTask class is introduced to handle
such tasks. This extension enables users to create tasks with start and
end dates, allowing for better organization of tasks that have time
constraints. Examples of such tasks include collecting certificates
between specific dates or attending events within a certain timeframe.

reetmitra committed:
Add enableAssertions to build.gradle file

reetmitra committed:
Merge branch 'branch-A-Assertions'

reetmitra committed:
Update README with EchoPilot commands and image embedding instructions

reetmitra committed:
Update README

reetmitra committed:
Update build.gradle to correct JAR file smoke test comptability

reetmitra committed:
Update build.gradle to correct JAR file smoke test comptability

reetmitra committed:
Update build.gradle to correct JAR file smoke test comptability

reetmitra committed:
Update README.md

reetmitra committed:
Update build.gradle to correct JAR file smoke test comptability

reetmitra committed:
Update build.gradle to correct JAR file smoke test comptability

reetmitra committed:
Update build.gradle to correct JAR file smoke test comptability

reetmitra committed:
Update build.gradle to correct JAR file smoke test comptability

reetmitra committed:
Update command parsing and saving task info

reetmitra committed:
Update command parsing and saving task info

reetmitra committed:
Change src files

reetmitra committed:
Update README.md

reetmitra committed:
Update to fix deadline command parsing issue

reetmitra committed:
Update to fix command parsing issue

reetmitra committed:
Update README.md
